keps/sig-api-machinery/customresources-pruning-and-defaulting.md

```diff
@@ -0,0 +1,486 @@
+---
+kep-number: 18
+title: Defaulting and Pruning for Custom Resources
+status: provisional
+authors:
+  - "@sttts"
+owning-sig: sig-api-machinery
+participating-sigs:
+  - sig-api-machinery
+  - sig-architecture
+reviewers:
+  - "@deads2k"
+  - "@lavalamp"
+  - "@liggitt"
+  - "@erictune"
+  - "@mbohlool"
+  - "@apelisse"
+approvers:
+  - "@deads2k"
+  - "@lavalamp - substituted by @fedebongio while @lavalamp is out"
+editor:
+  name: "@sttts"
+creation-date: 2018-07-31
+last-updated: 2018-07-31
+---
+
+# Defaulting and Pruning for Custom Resources
+    
+## Table of Contents
+* [Defaulting and Pruning for Custom Resources](#defaulting-and-pruning-for-custom-resources)
+    * [Table of Contents](#table-of-contents)
+    * [Overview](#overview)
+    * [Goals](#goals)
+    * [Non-Goals](#non-goals)
+    * [Motivation](#motivation)
+        * [Pruning](#pruning)
+        * [Defaulting](#defaulting)
+        * [Mixing of Schema and Value Validation](#mixing-schema-and-value-validation)
+    * [Formal Proposal – following pruning option 1]()
+        * [Types and Formats](#types-and-formats)
+        * [Polymorphic Fields](#polymorphic-fields)
+        * [Defaulting](#restrictions-on-defaults)
+        * [Excluding values from Pruning](#excluding-values-from-pruning)
+    * [Related topis](#related-topics)
+        * [Ratcheting (slightly related)](#ratcheting-(slightly-related))
+    * [References](#references)
+    * [Alternatives Considered](#alternatives-considered)
+
+## Overview
+
+Native Golang based resources do not persist JSON fields which are not part of the Golang structs which are backing them in the API server memory. This artifact of the use of typed Golang structs inside of the REST implementation has turned into an API convention. Major parts of the Kubernetes depend on this to ensure consistency of data persisted to etcd and returned from the REST API. 
+
+Without consistency of data in etcd, objects can suddenly render unaccessible on version upgrade because unexpected data may break decoding. Even if persisted data has correct format and decodes correctly, having it not gone through validation and admission when it was stored can break cluster-wide invariants. For example assume the `privileged: true/false` field is added to a type in Kubernetes version X+1. In version X, there is no security check around this. So every user could set that flag if we didn’t drop unknown fields. When the field is added in X+1, that user suddenly has escalated access (note: on read we do not run admission). This is a serious security risk.
```

Comment:
> I can see that bad data might break decoding for types with generated decoders that make assumptions, such as that a property value is numeric not string. But, would this happen for CRD decoding? Since they decode to Unstructured, there presumably aren't the same assumptions.

CRDs are often paired with clients generated from strongly typed structs. Those clients would fail to get/list/watch if the API server returned schema-invalid data. xref https://github.com/kubernetes/kubernetes/issues/57705

Reply:
@erictune I had those clients in mind as @liggitt writes. I will clarify the text.

Reply:
done